File: src/jray/math/Triangle.java
package jray.math;

import jray.common.Vect3;

public class Triangle {
	public static Vect3 getCircumScribedCircleCenter(Vect3 a, Vect3 b, Vect3 c){
		Vect3 mab = new Vect3();
		Vect3 mac = new Vect3();
		Vect3 normal = new Vect3();
		Vect3 dab = new Vect3();
		Vect3 dac = new Vect3();
		
		Vect.subtract(b, a, mab);
		Vect.subtract(c, a, mac);
		Vect.crossProduct(mab, mac, normal);
		Vect.crossProduct(mab, normal, dab);
		Vect.crossProduct(mac, normal, dac);
		Vect.addMultiple(a, mab, 0.5, mab);
		Vect.addMultiple(a, mac, 0.5, mac);
		dab.normalize();
		dac.normalize();
		
		if(Math.abs(dac.data[0])>1e-9&&Math.abs(dac.data[1])>1e-9&&Math.abs(dab.data[0]/dac.data[0] - dab.data[1]/dac.data[1])>1e-9){
			double x = ((mab.data[1]-mac.data[1])/dac.data[1] - (mab.data[0]-mac.data[0])/dac.data[0])/(dab.data[0]/dac.data[0] - dab.data[1]/dac.data[1]);

			Vect.addMultiple(mab, dab, x);
			return mab;
		}else if(Math.abs(dac.data[0])>1e-9&&Math.abs(dac.data[2])>1e-9&&Math.abs(dab.data[2]/dac.data[2] - dab.data[0]/dac.data[0])>1e-9){
			double x = ((mab.data[0]-mac.data[0])/dac.data[0] - (mab.data[2]-mac.data[2])/dac.data[2])/(dab.data[2]/dac.data[2] - dab.data[0]/dac.data[0]);
			
			Vect.addMultiple(mab, dab, x);
			return mab;
		}else if(Math.abs(dac.data[1])>1e-9&&Math.abs(dac.data[2])>1e-9&&Math.abs(dab.data[1]/dac.data[1] - dab.data[2]/dac.data[2])>1e-9){
			double x = ((mab.data[2]-mac.data[2])/dac.data[2] - (mab.data[1]-mac.data[1])/dac.data[1])/(dab.data[1]/dac.data[1] - dab.data[2]/dac.data[2]);
			
			Vect.addMultiple(mab, dab, x);
			return mab;
		}else if(Math.abs(dab.data[1])>1e-9&&Math.abs(dab.data[0])>1e-9&&Math.abs(dac.data[0]/dab.data[0] - dac.data[1]/dab.data[1])>1e-9){
			double y = ((mac.data[1]-mab.data[1])/dab.data[1] - (mac.data[0]-mab.data[0])/dab.data[0])/(dac.data[0]/dab.data[0] - dac.data[1]/dab.data[1]);
			
			Vect.addMultiple(mac, dac, y);
			return mac;
		}else if(Math.abs(dab.data[2])>1e-9&&Math.abs(dab.data[0])>1e-9&&Math.abs(dac.data[2]/dab.data[2] - dac.data[0]/dab.data[0])>1e-9){
			double y = ((mac.data[0]-mab.data[0])/dab.data[0] - (mac.data[2]-mab.data[2])/dab.data[2])/(dac.data[2]/dab.data[2] - dac.data[0]/dab.data[0]);
			
			Vect.addMultiple(mac, dac, y);
			return mac;
		}else if(Math.abs(dab.data[2])>1e-9&&Math.abs(dab.data[1])>1e-9&&Math.abs(dac.data[1]/dab.data[1] - dac.data[2]/dab.data[2])>1e-9){
			double y = ((mac.data[2]-mab.data[2])/dab.data[2] - (mac.data[1]-mab.data[1])/dab.data[1])/(dac.data[1]/dab.data[1] - dac.data[2]/dab.data[2]);
			
			Vect.addMultiple(mac, dac, y);
			return mac;
		}else{
			throw new RuntimeException("implement more cases...");
		}
	}
}

File: JRayXLib/Math/Triangle.cs
using System;
using JRayXLib.Shapes;

namespace JRayXLib.Math
{
    public class Triangle
    {
        public static Vect3 GetCircumScribedCircleCenter(Vect3 a, Vect3 b, Vect3 c)
        {
            Vect3 mab = b - a;
            Vect3 mac = c - a;
            Vect3 normal = mab.CrossProduct(mac);
            Vect3 dab = mab.CrossProduct(normal);
            Vect3 dac = mac.CrossProduct(normal);
            mab = a + mab/2;
            mac = a + mac/2;
            dab = dab.Normalize();
            dac = dac.Normalize();


            if (System.Math.Abs(dac.X) > Constants.EPS &&
                System.Math.Abs(dac.Y) > Constants.EPS &&
                System.Math.Abs(dab.X/dac.X - dab.Y/dac.Y) > Constants.EPS)
            {
                double x = ((mab.Y - mac.Y)/dac.Y - (mab.X - mac.X)/dac.X)/(dab.X/dac.X - dab.Y/dac.Y);
                return mab + dab*x;
            }

            if (System.Math.Abs(dac.X) > Constants.EPS &&
                System.Math.Abs(dac.Z) > Constants.EPS &&
                System.Math.Abs(dab.Z/dac.Z - dab.X/dac.X) > Constants.EPS)
            {
                double x = ((mab.X - mac.X)/dac.X - (mab.Z - mac.Z)/dac.Z)/(dab.Z/dac.Z - dab.X/dac.X);
                return mab + dab*x;
            }

            if (System.Math.Abs(dac.Y) > Constants.EPS &&
                System.Math.Abs(dac.Z) > Constants.EPS &&
                System.Math.Abs(dab.Y/dac.Y - dab.Z/dac.Z) > Constants.EPS)
            {
                double x = ((mab.Z - mac.Z)/dac.Z - (mab.Y - mac.Y)/dac.Y)/(dab.Y/dac.Y - dab.Z/dac.Z);
                return mab + dab*x;
            }

            if (System.Math.Abs(dab.Y) > Constants.EPS &&
                System.Math.Abs(dab.X) > Constants.EPS &&
                System.Math.Abs(dac.X/dab.X - dac.Y/dab.Y) > Constants.EPS)
            {
                double y = ((mac.Y - mab.Y)/dab.Y - (mac.X - mab.X)/dab.X)/(dac.X/dab.X - dac.Y/dab.Y);
                return mac + dac*y;
            }

            if (System.Math.Abs(dab.Z) > Constants.EPS &&
                System.Math.Abs(dab.X) > Constants.EPS &&
                System.Math.Abs(dac.Z/dab.Z - dac.X/dab.X) > Constants.EPS)
            {
                double y = ((mac.X - mab.X)/dab.X - (mac.Z - mab.Z)/dab.Z)/(dac.Z/dab.Z - dac.X/dab.X);
                return mac + dac*y;
            }

            if (System.Math.Abs(dab.Z) > Constants.EPS &&
                System.Math.Abs(dab.Y) > Constants.EPS &&
                System.Math.Abs(dac.Y/dab.Y - dac.Z/dab.Z) > Constants.EPS)
            {
                double y = ((mac.Z - mab.Z)/dab.Z - (mac.Y - mab.Y)/dab.Y)/(dac.Y/dab.Y - dac.Z/dab.Z);
                return mac + dac*y;
            }

            throw new Exception("implement more cases...");
        }
    }
}